File: java/search_in_rotated_sorted_array.java
public class Solution {
    public int search(int[] nums, int target) {
        int n=nums.length;
        if (n==0) return -1;
        int l=0, r=n-1;
        
        while (l<=r) {
            int mid=l+(r-l)/2;
            if (nums[mid]==target) return mid;
            if (nums[mid]<nums[r]) {
                if (target>nums[mid] && target<=nums[r]) {
                    l=mid+1;
                } else {
                    r=mid-1;
                }
            } else {
                if (target>=nums[l] && target<nums[mid]) {
                    r=mid-1;
                } else {
                    l=mid+1;
                }
            }
        }
        return -1;
    }
}

File: python/search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l, r=0, len(nums)-1
        while l<=r:
            mid=int(l+(r-l)/2)
            if nums[mid]==target: return mid
            else:
                if nums[mid]<nums[r]:
                    if target>nums[mid] and target<=nums[r]: l=mid+1
                    else: r=mid-1
                else:
                    if target<nums[mid] and target>=nums[l]: r=mid-1
                    else: l=mid+1
            
        return -1
